### PR TITLE
DAOS-10473 control: Do not require IOMMU for emulated NVMe

### PR DESCRIPTION
**Pull request by tanabarr**
Apply the following NVMe emulation related changes:

* If AIO is being used for NVMe emulation, don't require IOMMU and/or
  VFIO to be enabled.
* Disallow the use of emulated and non-emulated NVMe in the same engine
  config.
* Disable VMD if NVMe emulation is being used.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

**Review by mjmac — COMMENTED** [review]
Code changes seem fine, but there seems to be a significant amount of new/changed logic without any accompanying test coverage. Just concerned that we may continue to see regressions unless we add tests for this use case.

File: src/control/server/server_utils.go [hunk 1]
```diff
@@ -179,26 +175,27 @@ func getEngineNUMANodes(log logging.Logger, engineCfgs []*engine.Config) []strin
...
 }
 
+// Prepare bdev storage. Assumes validation has already been performed on server config. Hugepages
+// are required for both emulated (AIO devices) and real NVMe bdevs. VFIO and IOMMU are not
+// required for emulated file launchedNVMe.
 func prepBdevStorage(srv *server, iommuEnabled bool) error {
 	defer srv.logDuration(track("time to prepare bdev storage"))
 
-	hasBdevs := cfgHasBdevs(srv.cfg)
+	if srv.cfg.DisableHugepages {
```
Comment (kjacque) [inline src/control/server/server_utils.go:184]: If there are bdevs defined, shouldn't we be returning an error here?

**Review by kjacque — CHANGES_REQUESTED** [review]

File: src/control/server/server_utils.go (hunk 1 above)
Comment (tanabarr) [reply src/control/server/server_utils.go:184]: That validation will already have been performed on the config, don't think we need to validate here as well. I can check here as well if you feel strongly but that case (disable_hugepages set when bdevs exist) is tested in TestServer_prepBdevStorage which runs config validation before calling prepBdevStorage()

**Review by kjacque — DISMISSED** [review]

**Review by knard-intel — APPROVED** [review]
OK for me

**Review by Michael-Hennecke — COMMENTED** [review]
see my comment on VMD ... not sure if this is the right place to check for that but it's probably a good idea to do that check somewhere

File: src/control/server/server_utils.go [hunk 2]
```diff
@@ -199,26 +195,27 @@ func getEngineNUMANodes(log logging.Logger, engineCfgs []*engine.Config) []strin
 	return nodes
 }
 
+// Prepare bdev storage. Assumes validation has already been performed on server config. Hugepages
+// are required for both emulated (AIO devices) and real NVMe bdevs. VFIO and IOMMU are not
+// required for emulated file launchedNVMe.
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/server_utils.go:200]: done

File: src/control/server/server_utils_test.go [hunk 3]
```diff
@@ -529,6 +530,21 @@ func TestServer_prepBdevStorage(t *testing.T) {
 			expMemSize:      16384,
 			expHugePageSize: 2,
 		},
+		// VMD not enabled in prepare request.
+		"non-nvme bdevs; vmd enabled": {
```
Comment (kjacque) [inline src/control/server/server_utils_test.go:534]: comment seems to contradict with the description here. Which is right?

**Review by kjacque — APPROVED** [review]
OK with landing as-is since my comment is trivial.

File: src/control/server/server_utils.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (knard-intel) [reply src/control/server/server_utils.go:200]: What does mean launchedNVME ?
Is it a typo error ?

**Review by knard-intel — APPROVED** [review]
OK for me.
Just one minor remark, but probably some misunderstanding from my side

**Comment by tanabarr** [comment]
> OK for me. Just one minor remark, but probably some misunderstanding from my side

I will correct the "launchedNVME" in a subsequent doc-change PR to fix a couple of minor miscellaneous text issues.

**Pull request merged**